proposals/0002-spatial-navigation.md

```diff
@@ -0,0 +1,143 @@
+- Feature Name: Spatial Navigation
+- Start Date: 2023-09-12
+- PR: https://github.com/videojs/rfcs/pull/2
+
+# Summary
+[summary]: #summary
+
+This feature aims to improve user experience and accessibility on smartTV devices by enabling seamless navigation through interactive elements within the Video.js player using remote control arrow keys.
+
+# Motivation
+[motivation]: #motivation
+
+Smart TV devices commonly use directional input, such as arrow keys on a remote control. By implementing spatial navigation using the spatial navigation polyfill, we aim to ensure that users can intuitively navigate through player controls and elements on smartTV devices.
+
+### Goals
+- Enable seamless and intuitive navigation within the Video.js player on smart TV platforms.
+- Create a new SpatialNavigation component.
+- Implement the spatial navigation polyfill.
+- Redesign the Video.js user interface to suit smart TV platforms and facilitate spatial navigation.
+
+### Non-Goals
+- Keycode mapping for smartTV devices that use non-standardized keycodes. The spatial navigation polyfill uses 37: 'left', 38: 'up', 39: 'right', and 40: 'down'.
+
+# Guide-level Explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Smart TV users will be able to navigate through Video.js player controls using arrow keys on their remote controls. The player will have an initial focus; when a user presses, for example, the "right" arrow key, the focus will shift to the next interactive control, allowing them to easily navigate through player UI elements. Focused UI elements will be selectable using the ENTER/OK button on the RCU.
+
+In order to allow the user to navigate out of the progress bar (Slider), the UP and DOWN buttons will not call the stepForward and stepBack functions, as is now. The LEFT and RIGHT buttons will retain the same functions but only when the progress bar (Slider) is in focus.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Dependency Management and Configuration Options
+For this feature, additional dependencies will be added: spatial-navigation-polyfill and flat-polyfill. Flat-polyfill is needed to accommodate devices lacking support for Array.prototype.flat.
```

Comment:
In general, we have traditionally avoided bundling polyfills with Video.js. Instead, we've preferred to allow users to supply a polyfill of their choosing (if any). This was how we handled `Promise` in earlier versions of Video.js.

The reasoning behind this is that Video.js does not need to be in the business of changing native APIs implicitly.

I'd suggest that Video.js can feature detect the spatial navigation API(s) before trying to do anything with them. Methods that depend on native spatial navigation (or a polyfill) would simply become no-ops when that support does not exist.

Reply:
Another reason not to include the polyfill by default is filesize.
The polyfill is 72K ungzipped and 14K gzipped. (It could be a bit smaller if also minified). 14K isn't an insignificant amount to add.

(When minified, it goes down to 34K and 7K mingzipped)

----------------

> Methods that depend on native spatial navigation (or a polyfill) would simply become no-ops when that support does not exist.

👍🏻

Reply:
> Methods that depend on native spatial navigation (or a polyfill) would simply become no-ops when that support does not exist.

As far as I understand, there is no native spatial navigation yet. There are no platforms that officially implemented this spec: https://www.w3.org/TR/css-nav-1/

Please correct me if I am wrong!

This means that users use this polyfill, other libraries, or a custom implementation for spatial navigation.

I think if we want to provide an out-of-the-box experience, we may consider this polyfill as a library, meaning we do not pollute global scope but rather import specific functions in our code base and use them.

Or we can provide a mechanism to set spatial navigation solution. This spatial navigation solution should implement a specific interface that we expect. In this scenario, users will have to implement adapters around their spatial navigation solution to conform interface that we expect.

Reply:
Thanks for the feedback. I understand your concerns about adding an additional polyfill. I'll remove all references to it.
@dzianis-dashkevich that's correct. Currently, there is no native spatial navigation on Samsung Tizen and LG webOS platforms.

Reply:
@misteroneill 
Does it mean you think that it will be better to go with something like option 2 from my message above?

Reply:
@dzianis-dashkevich Yes, I think something like option 2 would be best.

We can model the interface we expect on the standard interface (if such exists) similar to the way Video.js techs model their interface on the `HTMLMediaElement` interface.